### ISSUE TITLE
mult_season_adjust logic

### ISSUE THREAD

**Issue by jsenergy** [cmdty/curves#15]
I am starting with a simple exercise of using the mult_season_adjust logic in max_smooth_interp to shape quarters within a year. This works alright if using e.g. the below inputs:

```
def wd_weight(period):
    start = period.asfreq('B', 's')
    end = period.asfreq('B', 'e')
    return len(pd.period_range(start=start, end=end, freq='B'))
```

```
pk_cont = [(datetime.date(2022, 1, 3), datetime.date(2022, 3, 31), 71.81),
           (datetime.date(2022, 1, 3), datetime.date(2022, 12, 30), 58.0)]

q_peak_piecewise_curve, q_peak_bc_daily = bootstrap_contracts(pk_cont, freq='Q', average_weight=wd_weight)

pk_ratio_config = {1: 1.07015, 2: 0.79521, 3: 0.9248, 4: 1.20984}
def pk_mult_adjust(period):
    return pk_ratio_config[period.quarter]
q_peak_smooth_curve = max_smooth_interp(q_peak_bc_daily, average_weight=wd_weight, freq='Q', mult_season_adjust=pk_mult_adjust)
```

q_peak_piecewise_curve keeps Q1 and has a flat price structure in Q2, Q3, Q4 that meets zero-arbitrage conditions with the provided Cal value.
q_peak_smooth_curve has later adjusted the ratios based on the config to reflect the summer/winter structure.

However, if I use the below set of inputs (Q1, Q3 and Cal), mult_season_adjust does no longer do the trick:

```
pk_cont = [(datetime.date(2022, 1, 3), datetime.date(2022, 3, 31), 71.81),
           (datetime.date(2022, 7, 1), datetime.date(2022, 9, 30), 45.73),
           (datetime.date(2022, 1, 3), datetime.date(2022, 12, 30), 58.0)]
```

q_peak_piecewise_curve keeps Q1 and Q3 values and has a flat price structure in Q2 and Q4 that meets zero-arbitrage conditions with the provided Cal value.
q_peak_smooth_curve no longer manages to cope with any ratio adjustments, meaning that smooth_curve = piecewise_curve. By specifying some contracts we by definition break the  pk_ratio_config ratios, however, we should be able to preserve some and leave the 'last' (to be specified somehow) contract to 'break' those ratios.

The next steps for me are to then go from quarterly to monthly, to weekly, to daily etc. Therefore, critical to get it right from the beginning.

Please let me know your thoughts.

**Comment by cmdty** [comment]
Yes, there is always the problem of over-specifying the problem by introducing redundancy between contracts and ratios. The bootstrap_contracts method does have the optional parameter "allow_redundancy" which means an exception isn't thrown in redundant cases, although I'm not sure whether I would recommend to ever set this to true. In your case definitely do not as the prices should take precedence over the ratios as they are real trades prices, rather than the ratios which I assume are estimated.

The challenge is to pick the correct inputs to feed into the bootstrapper, and which ones to discard. I did think about optionally including this logic in the bootstrap_contracts function but decided against it. Implementing this logic in a robust manner is very difficult considering the algorithm internals don't know what sort of inputs are going to be passed in. Maybe there will be gaps in the input contract, a varied combination of contract granularities (i.e. daily, weekend, weekly, monthly prices etc), which would make it very difficult to programmatically decide which input data to discard and which to keep. Seeing two monthly input contracts and an associated ratio it is easy to decide to discard the ratio. However, this can get much more complex. For example you could have a day-ahead price, a balance of month price (both of which cover a whole month) the next monthly contract, and the ratio between these contracts. In this example the ratio is also redundant and should be discarded, but you can see the logic getting harder, and it could get much harder than this.

This code would probably end up being more complex than the actual core bootstrapping and much less robust. So I decided to leave the bootstrapper as a mathematical routine leaving it up to the calling code to choose the inputs, which is a better place for it. If you are writing a tool using the bootstrapper, the tool author will have much better information about the structure of the data which will be going to be fed into bootstrap_contracts and thus how to filter appropriately and robustly.

**Comment by jsenergy** [comment]
I understand that there is a potential redundancy between contracts and ratios. 

However, the case of contract/ratio redundancy is equivalent in a Q1/Q2/Cal and Q1/Q3/Cal scenario. Your LP model solves for the former without issue, but the latter somehow struggles.

I am trying to design the inputs in the most simple way - these are only quarters vs Cal. How would you suggest I structure the inputs for it to work and how should one think about more granular structures?

**Comment by cmdty** [comment]
In the example you gave which gives a bad result (Q1/Q3/Cal) including a ratio or spread, either Q1-Q2 or Q2-Q3 should solve this issue. Q4 will be boostrapped to the arb free price. So if you've got gaps in contracts of the same granularity you can fill them in with ratios or spreads. The difficulty is doing this without introducing redundancy, for example if you have Q1/Q3/Summer then adding shaping spread or ratio will introduce redundancy as the Q2 price is already fully specified by Q3 and Summer prices. I guess this is another example of how programmatically choosing the right inputs without introducing redundancy could be quite difficult.

One way to make the program simpler is to enforce that all contracts of the same granularity are contiguous, so there are not gaps. You then also need to discard contracts where those of a higher granularity completely specify a contract. Have contiguous contracts will also make this discard logic also simpler.

**Comment by jsenergy** [comment]
I will meed to have a think about the most sensible approach. Removing
non-contiguous contracts is probably the best starting point.

Let me experiment a little with contract non-contiguity and see what I can
d.

The issue is that sometimes you have pretty good transparency on deferred
winter contracts while summer remains without a defined structure. In an
ideal scenario, those Q4 values would be used, while missing ones would
default to the the mult_season_adjust ratios.

Another option would be to stop using the mult_seasons_adjustargument and
structure every contract relationship using shaping_ratios in
bootstrap_contracts instead.

Thanks for your guidelines.

**Comment by cmdty** [comment]
Deciding between using shaping ratios/spreads in the bootstrapper and the adjustments in the spline is a tradeoff:

**Pros of using ratios/spreads in bootstrapper:**
- Transparent and easy to understand.
- Ratios/spreads can be estimated from historical data and used in a formulaic curve building strategy, something which is probably necessary if building middle office or risk department curves.

**Cons of using ratios/spreads in bootstrapper:**
- If the market prices move a large amount, applying the same shaping in the bootstrapper can result in curves which just don't look right, e.g. you might end up in a cliff-edge difference in prices between adjacent contracts such as Dec and Jan.

**Pros of using spline adjustment:**
- Will often result in a more pleasing looking smooth curve.

**Cons of using spline adjustment:**
- Someone arbitrary and I'm sure if there is a formulaic way to estimate the adjustment. This could make it a no-go for middle office curve construction.
